wire-runtime/src/main/java/com/squareup/wire/MessageAdapter.java

```diff
@@ -534,6 +558,29 @@ private void writeValueNoTag(WireOutput output, Object value, Datatype datatype)
 
   // Reading
 
+  /** Read an encoded message from {@code source}. */
+  public M read(BufferedSource source) throws IOException {
+    checkNotNull(source, "source == null");
+    return read(new WireInput(source));
+  }
+
+  /** Read an encoded message from {@code bytes}. */
+  public M readBytes(byte[] bytes) throws IOException {
+    checkNotNull(bytes, "bytes == null");
+    Buffer buffer = new Buffer();
+    M message = read(buffer.write(bytes));
+    if (!buffer.exhausted()) {
+      throw new IllegalArgumentException("");
+    }
+    return message;
+  }
+
+  /** Read an encoded message from {@code stream}. */
+  public M readStream(InputStream stream) throws IOException {
```

Comment:
This method is slightly hazardous, in that it can read more bytes than it's supposed to. Imagine a stream that has protocol buffers stacked end-on-end. Sigh.